### PR TITLE
Use Typescript and quicktype-wrapper for code generation

### PR DESCRIPTION
**Pull request by michaelawyu**
~~To-Do: Update the GitHub repository to the official repo when [PR #70](https://github.com/googleapis/google-cloudevents/pull/70) is merged.~~

### CODE REVIEW

File: gen/templates/README.squirrelly [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+go install github.com/googleapis/google-cloudevents-go
+```
+
+To use an event class, see the snippet below:
+
+``` go
+package main
+
+import (
+	pubsub "github.com/googleapis/google-cloudevents-go/cloud/pubsub/v1"
```
Comment (grant) [inline gen/templates/README.squirrelly:28]: Spacing issue?

File: gen/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  if (!IN || !OUT) return;
+  if (IN.endsWith('/')) IN = IN.substring(0, IN.length - 1);
+  if (OUT.endsWith('/')) OUT = OUT.substring(0, OUT.length - 1);
+
+  const templateDirectoryPath = `${__dirname}/../../${TEMPLATE_DIRECTORY}`;
+  const snippetDirectoryPath = `${__dirname}/../../${SNIPPET_DIRECTORY}`;
+
+  const schemasAndGenFiles = await qt.getJSONSchemasAndGenFiles(IN, LANGUAGE);
+  const allEvents: Array<[string, string, string]> = [];
+  schemasAndGenFiles.map((sg: [any, string]) => {
```
Comment (grant) [inline gen/src/index.ts:32]: A few more comments in this whole file would be helpful.

File: gen/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    pkg = pkg.replace('google.events.', '');
+    const pkgPath = pkg.replace(/\./g, '/');
+    mkdirp.sync(`${OUT}/${pkgPath}`);
+
+    // Update the package statement
+    const goPkg = schema['goPackage'];
+    const goPkgStmt = `package ${goPkg}`;
+    genFile = genFile.replace('package main', goPkgStmt);
+
+    const pkgEvents: Array<string> = [];
```
Comment (grant) [inline gen/src/index.ts:46]: `Array<string>` should be `string[]`.

File: gen/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    Object.keys(schema['properties']).map((eventName: string) => {
+      const eventDescription = schema['properties'][eventName]['description'];
```
Comment (grant) [inline gen/src/index.ts:48]: Use direct property access where possible:

```ts
const eventDescription = schema.properties[eventName].description;
```

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
+FROM node:12
```
Comment (grant) [inline Dockerfile:1]: This file should not be in the root folder as it is only used for generating this library. Maybe a folder like `gen`?

**Review by grant — CHANGES_REQUESTED** [review]
Looks good overall.

The main thing is:
- Can we have a working samples i.e. for Pub/Sub events? That event is the most important event type we are generating.

Overall, what do you think of this approach rather than the protobuf dependency way? From the surface, it looks like it's better?

File: Dockerfile (hunk 5 above)
Comment (michaelawyu) [reply Dockerfile]: Done.

File: gen/src/index.ts (hunk 4 above)
Comment (michaelawyu) [reply gen/src/index.ts]: Done.

File: gen/src/index.ts (hunk 3 above)
Comment (michaelawyu) [reply gen/src/index.ts]: Done.

File: gen/src/index.ts (hunk 2 above)
Comment (michaelawyu) [reply gen/src/index.ts:34]: Done.

File: gen/templates/README.squirrelly (hunk 1 above)
Comment (michaelawyu) [reply gen/templates/README.squirrelly]: Done.

File: README.md [hunk 6]
```diff
@@ -1,19 +1,45 @@
...
+
+func main () {
+    e := audit.UnmarshalMessagePublishedEvent(data)
+    fmt.Println(e.message)
+}
 ```
 
-Observe the new files in `src/`.
\ No newline at end of file
+## Generation
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply README.md]: Done.

File: README.md [hunk 7]
```diff
@@ -1,19 +1,45 @@
...
+package main
+
+import (
+	pubsub "github.com/googleapis/google-cloudevents-go/cloud/pubsub/v1"
+)
+
+func main () {
+    e := audit.UnmarshalMessagePublishedEvent(data)
+    fmt.Println(e.message)
+}
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply README.md]: Will send in a separate PR.

**Review by grant — APPROVED** [review]

**Pull request merged**